base/utfcheck.jl

```diff
@@ -0,0 +1,255 @@
+                # check for surrogate pairs, make sure correct
+                if is_surrogate_codepoint(ch)
+                    !is_surrogate_lead(ch) && utf_errfunc(UTF_ERR_NOT_LEAD, pos-2, ch)
+                    # next character *must* be a trailing surrogate character
+                    (pos + 3 > len) && utf_errfunc(UTF_ERR_MISSING_SURROGATE, pos-2, ch)
+                    byt = dat[pos += 1]
+                    (byt != 0xed) && utf_errfunc(UTF_ERR_NOT_TRAIL, pos, byt)
+                    surr = get_continuation(0x0000d, dat[pos += 1], pos)
+                    surr = get_continuation(surr, dat[pos += 1], pos)
+                    !is_surrogate_trail(surr) && utf_errfunc(UTF_ERR_NOT_TRAIL, pos-2, surr)
+                    (options & UTF_NO_SURROGATES) != 0 && utf_errfunc(UTF_ERR_SURROGATE, pos-2, surr)
+                    flags |= UTF_SURROGATE
+                    num4byte += 1
+                elseif ch > 0x07ff
+                    num3byte += 1
+                elseif (options & UTF_ACCEPT_LONG) != 0
+                    flags |= UTF_LONG
+                    num2byte += 1
+                else
+                    utf_errfunc(UTF_ERR_LONG, pos-2, ch)
+                end
+            elseif ch < 0xf5
+                # 4-byte UTF-8 sequence (i.e. characters > 0xffff)
+                (pos + 3 > len) && utf_errfunc(UTF_ERR_SHORT, pos, ch)
+                ch = get_continuation(ch & 0x07, dat[pos += 1], pos)
+                ch = get_continuation(ch, dat[pos += 1], pos)
+                ch = get_continuation(ch, dat[pos += 1], pos)
+                if ch > 0x10ffff
+                    utf_errfunc(UTF_ERR_INVALID, pos-3, ch)
+                elseif ch > 0xffff
+                    num4byte += 1
+                elseif is_surrogate_codepoint(ch)
+                    utf_errfunc(UTF_ERR_SURROGATE, pos-3, ch)
+                elseif (options & UTF_ACCEPT_LONG) != 0
+                    # This is an overly long encode character
+                    flags |= UTF_LONG
+                    if ch > 0x7ff
+                        num3byte += 1
+                    elseif ch > 0x7f
+                        num2byte += 1
+                    end
+                else
+                    utf_errfunc(UTF_ERR_LONG, pos-2, ch)
+                end
+            else
+                utf_errfunc(UTF_ERR_INVALID, pos, ch)
+            end
+        end
+    end
+    num3byte != 0 && (flags |= UTF_UNICODE3)
+    num4byte != 0 && (flags |= UTF_UNICODE4)
+    return totalchar, flags, num4byte, num3byte, num2byte
+end
+
+#=
+@doc """
+@brief      Validates and calculates number of characters in a UTF-16 encoded vector of UInt16
+
+@param[in]  dat     Vector{UInt16}
+@param[in]  options flags to determine error handling (default 0)
+
+@return     (total characters, flags, 4-byte, 3-byte, 2-byte)
+@throws     ArgumentError
+""" ->
+=#
+function check_string_utf16(dat::Vector{UInt16}, len::Int)
```

Comment:
>> One deals with UTF-8 encoding, the second with UTF-16 encoding, the third with UTF-32 encoding... and as far as merging check_string_utf32 and check_string_abs, somebody on this thread told me that I shouldn't mix Vector{Char} and AbstractString... (i.e. mix abstraction and implementation) in the same function.
> I did, but my objection is actually about check_string_abs existing at all. I still don't understand why it checks for surrogates, while getindex should only return valid codepoints, never surrogates. So I don't think you need to check for the validity of AbstractString in general: if you access them using getindex (or next), I think you should assume they are correct. It's much more reasonable to leave the checking to the specific AbstractString type, which has more information about what checks may be needed (for example, you might have a CheckedUTF8 or and UncheckedUTF8 custom type which would skip some checks).

I think the general tension here is that there are a small number of methods (3 actually) implemented for Vector types (but not consistently) which aren't really that sensible or . I think it would be better to drop the three `convert(::UTF8String, ::Vector{UInt16})`, `convert(::Union(UTF16String,UTF8String), ::Vector{UInt32}) methods and then all of these methods can work directly on string types instead of implicitly passing around the data vector portion only and trying to figure out (or pass along context, or duplicate code) to track whether the trailing null value was supposed to be part of the string. The definition of these methods are inconsistent with the definition of `convert(::Union(UTF16String, UTF32String), ::Vector{UInt8})` too, which I find confusing and thus especially sensible to remove. (yes, i realize all the method signatures I am referring to are pre-existing to your PR, but part of the point of this PR is to cleanup the behavior of this code, so I'm just reviewing all of it)